Enable unit tests and fix all the issues found thanks to them

mardy opened the pull request:
It might be simpler to review this commit by commit, as there are many unrelated changes. The unifying point is that all these commits are either related to enabling running unit tests, or to fixing errors found thanks to them.

mardy committed:
debian: Run tests from build dir

mardy committed:
Update for eds 3.18

mardy committed:
tests: remove failing comparison in collections test

The check doesn't match the comment text (seems to do exactly the
opposite!) and randomly fails even if the condition is inverted.

mardy committed:
tests: register metatype used in signal argument

mardy committed:
tests: fix timezone in recurrence-test

This changed with 4ca7dc97992e8eda9119a8075b82bed80b0409fc

mardy committed:
debian/rules: don't run tests if the build profile contains "nocheck"

mardy committed:
QtOrganizer: fix reading of recurrent events

mardy committed:
tests: update eds-base-test to work with the new ID scheme

mardy committed:
QtOrganizer: properly handle items with no collection assigned

This issue was discovered while running unit tests: Items which get
created with no collection assigned, should be bound to the default
collection.

mardy committed:
tests: Do not leak signals across test cases

When saving an item, the itemsAdded() signal will be emitted
asynchronosuly some time later. We need to wait for it, before ending
the test, otherwise it might alter the behaviour of the following test.

mardy committed:
tests: fix item ID comparison

The item IDs are now encoded using base64.

mardy committed:
tests: disable floating time test

Items with floating times are no longer supported: since commit
ac75731c1233fafcdfa9c01e5f580e048ff49cff we interpret an invalid
timezone as UTC, and moreover QDateTime is no longer working properly
when an invalid timezone is set.

mardy committed:
tests: connect to signal spies using C++11 syntax

mardy committed:
tests: fix double free error:

The icalendar object is already deleted by the list free operation one
line above.

hummlbach reviewed (commented):
I looked over the commits for half an hour, but i would need half an hour per commit (due to me lacking knowledge about qtpim, eds and the way unit tests are done here, inspite the changes being very small). So I concentrate on the two changes in the production code and pose a question on of them...

hummlbach commented on organizer/qorganizer-eds-saverequestdata.cpp:
```diff
@@ -33,6 +33,14 @@ SaveRequestData::SaveRequestData(QOrganizerEDSEngine *engine,
+             * associated collection; we cannot leave it to null here, as the
+             * caller of nextSourceId() would consider that as having reached
+             * the end of the list.
+             */
```
I'm not sure whether I agree on that comment...
After the call of `nextSourceId` at qorganizer-eds-engine.cpp:567 it is checked `sourceId` is null, but also if `data->end()` is true too `data->finish()` gets called.
But on the other hand `data->finish()` gets called at qorganizer-eds-engine.cpp:163 if `sourceId.isEmpty()`. Is that really consistent and correct?
Should the comment refer to qorganizer-eds-engine.cpp:585, where the default collection is assumed when the `sourceId` is empty? Could we theoretically change it to `isNull` there instead of setting it to `""` here?

mardy replied:
Thanks @hummlbach for the review! Yes, I agree that it's confusing, and I've added a comment at qorganizer-eds-engine.cpp:585, as you suggested. I think the other comment I've added before is still very important, so I've left it too.
The reason why in qorganizer-eds-engine.cpp:163 we use a different check is because in that situation we are *reading* data out of the databse, not *writing* it: and we know that all items stored into the database will have a non-null, non-empty collection ID assigned to them.

hummlbach replied:
I still don't get it completely... But don't mind :grin:

hummlbach commented:
Something else: I was about to click on the merge button, when I saw the collections test was failing for amd64 with the last build... To really make sense of the tests, we should fix that right...?

mardy commented:
You are right, and I'll probably remove that line from the test, since I don't understand what it does and the comment next to it makes it even more confusing. :-)
But I notice that there's another warning from another test case, and I'll definitely fix that for good :-)

mardy commented:
BTW, once I've fixed it, please just approve the MP, don't merge it: I prefer when the author merges the MP himself, to keep the responsibilities more evident (at least when the author has merge rights).

mardy committed:
Tests: remove debug message

mardy committed:
Tests: allow running the test under valgrind

mardy commented:
@hummlbach, I've updated the branch. Unfortunately, after many hours of debugging, I couldn't get around the unpredictability of the "readonly" property, so the tests are still randomly failing. But I think it's still good if we run them (I've modified the `debian/rules` file to ignore their exit status), and [I've filed a bug](https://gitlab.gnome.org/GNOME/evolution-data-server/issues/53) in the EDS project.

mardy committed:
QtOrganizer: fix an error reported by valgrind

The iterator becomes invalid once removed, se we must use it before
removing it :-)

mardy committed:
QtOrganizer: fix a race condition on collection creation

This was reported as an EDS bug here:
https://gitlab.gnome.org/GNOME/evolution-data-server/issues/53

The problem is taht the old code was assuming (and maybe this was
working fine with older EDS versions?) that the "source-added" signal
from the source registry would be emitted before the
e_source_registry_create_sources() method returned. But this is not
always the case.

mardy committed:
tests: Be flexible about the number of emitted signals

EDS might compress multiple invocations of the same signal into a single
one.
